hail/src/main/scala/is/hail/expr/ir/Parser.scala

```diff
@@ -0,0 +1,1009 @@
+package is.hail.expr.ir
+
+import is.hail.HailContext
+import is.hail.expr.{JSONAnnotationImpex, ParserUtils}
+import is.hail.expr.types._
+import is.hail.expr.types.physical.PType
+import is.hail.rvd.RVDType
+import is.hail.table.{Ascending, Descending, SortField}
+import is.hail.utils.StringEscapeUtils._
+import is.hail.utils._
+import is.hail.variant.ReferenceGenome
+import org.json4s.jackson.{JsonMethods, Serialization}
+
+import scala.util.parsing.combinator.JavaTokenParsers
+import scala.collection.JavaConverters._
+import scala.reflect.ClassTag
+import scala.util.parsing.input.Positional
+
+abstract class Token extends Positional {
+  def value: Any
+}
+
+final case class IdentifierToken(value: String) extends Token { }
+final case class StringToken(value: String) extends Token { }
+final case class IntegerToken(value: Long) extends Token { }
+final case class FloatToken(value: Double) extends Token { }
+final case class PunctuationToken(value: String) extends Token { }
+
+case class TokenIterator(a: Array[Token]) extends BufferedIterator[Token] {
+  private var pos = 0
+
+  def next(): Token = {
+    val x = a(pos)
+    pos += 1
+    x
+  }
+
+  def hasNext(): Boolean = pos < a.length
+
+  def head(): Token = {
+    assert(hasNext())
+    a(pos)
+  }
+}
+
+object IRLexer extends JavaTokenParsers {
+  val token: Parser[Token] =
+    "[()\\[\\]{}<>,:+@=]".r ^^ { p => PunctuationToken(p) } |
+      identifier ^^ { id => IdentifierToken(id) } |
+      float64_literal ^^ { d => FloatToken(d) } |
+      int64_literal ^^ { l => IntegerToken(l) } |
```

Comment:
I think int literal should before float literal so you don't mistakenly parse 5 as 5.0.

Reply:
If I do this, then I get this error:

```
is.hail.utils.HailException: Expected punctuation ')' but found float '0.14'.
<input>:1:(F32 3.14)
                           ^
```